Added modified API.md and new APISocketOptions.md (#1510)
Haivision/srt: docs/API.md

# SRT API

The SRT C API (defined in `srt.h` file) is largely based in design on the legacy 
UDT API, with some important changes. The `udt.h` file contains the legacy UDT API 
plus some minor optional functions that require the C++ standard library to be used. 
There are a few optional C++ API functions stored there, as there is no real C++ API 
for SRT. These functions may be useful in certain situations.

There are some example applications so that you can see how the API is being used, 
including `srt-live-transmit`, `srt-file-transmit` and `srt-multiplex`. All SRT 

> related material is contained in `transmitmedia.*` files in the `common` directory 

Reviewer: Things have changed over time here and it wasn't updated:

1. `srt-multiplex` no longer exists. The architecture was updated in `srt-live-transmit` and therefore this application has been moved to test, as only they preserve the old architecture. So this file should be removed from this list.
2. The directory that contains these `transmitmedia.*` files is now `apps`, not `common`.

Rest of the description can stay as is.

> There are some example applications so that you can see how the API is being used, 
> including `srt-live-transmit`, `srt-file-transmit` and `srt-multiplex`. All SRT 
> related material is contained in `transmitmedia.*` files in the `common` directory 

Author: ```suggestion
There are some example applications so that you can see how the API is being used, 
including `srt-live-transmit` and `srt-file-transmit`. All SRT related material is contained 
in `transmitmedia.*` files in the `apps` directory 
```

The **simple API** includes: `srt_send` and `srt_recv` functions. They need only
the socket and the buffer to send from or receive to, just like system `read`
and `write` functions.

The **rich API** includes the `srt_sendmsg` and `srt_recvmsg` functions. Actually
`srt_recvmsg` is provided for convenience and backward compatibility, as it is
identical to `srt_recv`. The `srt_sendmsg` receives more parameters, specifically
for messages. The `srt_sendmsg2` and `srt_recvmsg2` functions receive the socket, 
buffer, and the `SRT_MSGCTRL` object, which is an input-output object specifying 
extra data for the operation.

Functions with the `msg2` suffix use the `SRT_MSGCTRL` object, and have the
following interpretation (except `flags` and `boundary` which are reserved for
future use and should be 0):

* `srt_sendmsg2`:
    * `msgttl`: [IN] maximum time (in ms) to wait for successful delivery (-1: indefinitely)
    * `inorder`: [IN] if false, the later sent message is allowed to be delivered earlier
    * `srctime`: [IN] timestamp to be used for sending (0 if current time)
    * `pktseq`: unused
    * `msgno`: [OUT] message number assigned to the currently sent message

* `srt_recvmsg2`

>     * `msgttl`: inorder; unused

Reviewer: Here were two fields mentioned in this entry. So, either leave as is, or split into two different enums.

>     * `msgttl`: inorder; unused

Author: ```suggestion
    * `msgttl`:  unused
    * `inorder`: unused
```

Transmission types available in SRT
-----------------------------------

Mode settings determine how the sender and receiver functions work. The main 

> [socket options](#https://github.com/Haivision/srt/blob/master/docs/API-SocketOptions.md) 

Reviewer: Dash in the name causes likely a wrong link.

OTOH I personally prefer the `API-SocketOptions.md` filename, you may consider renaming it.

BTW. I think it will be better if the links to the document use a relative, not network-absolute path.

> [socket options](#https://github.com/Haivision/srt/blob/master/docs/API-SocketOptions.md) 

Author: ```suggestion
[socket options](APISocketOptions.md) 
```

SRT functions can also work in blocking and non-blocking mode, for which
there are two separate options for sending and receiving: `SRTO_SNDSYN` and
`SRTO_RCVSYN`. When blocking mode is used, a function will not exit until
the availability condition is satisfied. In non-blocking mode the function 
always exits immediately, and in case of lack of resource availability, it
returns an error with an appropriate code. The use of non-blocking mode usually
requires using some polling mechanism, which in SRT is **EPoll**.

Note also that the blocking and non-blocking modes apply not only for sending
and receiving. For example, `SNDSYN` defines blocking for `srt_connect` and

> `RCVSYN` defines blocking for `srt_accept`. The `SNDSYN` also makes `srt_close`

Reviewer: That should be updated as this is completely wrong.

The blocking for `srt_connect` is set by setting `SRTO_RCVSYN` to false (surprisingly, as the connection-ready state is `SRT_EPOLL_OUT`). The blocking for `srt_accept` is, also surprisingly, set by `SRTO_RCVSYN` on the listener socket from which the accepting is done, although the readiness for acceptance is marked by `SRT_EPOLL_IN` event.

Since the second sentence, I'd add rather something like this:

To make `srt_connect` running asynchronously, you should set the `SRTO_RCVSYN` to false on the socket used for connecting (the `SRT_EPOLL_OUT` readiness should be checked here). To make the `srt_accept` running asynchronously, you should set `SRTO_RCVSYN` to false on the listener socket (the `SRT_EPOLL_IN` readiness should be checked here). The `SRTO_SNDSYN` set to true on the socket makes also `srt_close` exit only after the sending buffer is completely empty.

Reviewer: (Updated the comment above - I messed up a bit with explanations, but the text in the last paragraph is ok.)

Reviewer: First of all, to make you imagine how it works: The `SRTO_SNDSYN` and `SRTO_RCVSYN` options are set to `true` by default, and this means **blocking** mode. So you should set appropriate option to `false` in order to have **non-blocking** mode. So, among things mentioned here:
* `SRTO_RCVSYN` set to `false` makes `srt_connect` run asynchronously (non-blocking)
* IDEM (on the listener socket), makes `srt_accept` run asynchronously (non-blocking)
* When `SRTO_SNDSYN` is true (default!) then also `srt_close` will block until the sender buffer is empty. When this is set to false, the `srt_close` function exits immediately, but sending will continue in the background.

Author: Blocking and Non-blocking Mode
==============================

SRT functions can also work in blocking and non-blocking modes, for which
there are two separate options for sending and receiving: `SRTO_SNDSYN` and
`SRTO_RCVSYN` (both are set to true by default, corresponding to blocking mode). 
When blocking mode is used, a function will not exit until the availability 
condition is satisfied. In non-blocking mode the function always exits immediately, 
and if enough resources are not available it returns an error with an appropriate 
code. The use of non-blocking mode usually requires using some polling mechanism, 
which in SRT is **EPoll**.

Note also that blocking and non-blocking modes apply not only to sending and 
receiving. For example, to make `srt_connect` run asynchronously (non-blocking mode), 
set `SRTO_RCVSYN` to false on the socket used for connecting (the `SRT_EPOLL_OUT` 
readiness should be checked here). To make `srt_accept` run asynchronously (non-blocking 
mode), set `SRTO_RCVSYN` to false on the listener socket (the `SRT_EPOLL_IN` readiness 
should be checked here). When `SRTO_SNDSYN` is true (default) then also `srt_close` 
will block until the sender buffer is empty. When `SRTO_SNDSYN` is set to false, the 
`srt_close` function exits immediately, but sending will continue in the background.

Reviewer: Perfect.

the original UDT library supported, that is, *file* and *message* transmission.

There are two general modes: **Live** and **File** transmission. Inside File 
transmission mode, there are also two possibilities: **Buffer API** and 
**Message API**. The Live mode uses Message API. However it doesn't exactly match 
the description of the Message API because it uses a maximum single sending buffer 
up to the size that fits in one UDP packet.

There are two options to set a particular type:

* `SRTO_TRANSTYPE`: uses the enum value with `SRTT_LIVE` for live mode
   and `SRTT_FILE` for file mode. This option actually changes several parameters 
   to their default values for that mode. After this is done, additional parameters, 
   including those that are set here, can be further changed.
* `SRTO_MESSAGEAPI`: This sets the Message API (true) or Buffer API (false)

This makes possible a total of three data transmission methods:

* [Live](#transmission-method-live)
* [Buffer](#transmission-method-buffer)
* [Message](#transmission-method-message)

### Terminology

The following terms are used in the description of transmission types:

> **HANGUP / RESUME**: When a function "HANGS UP" this means that it returns an 

Reviewer: Actually the intention of this description was to have some common abstract terms for what happens differently depending on whether the function works in blocking and non-blocking mode.

If the function makes a "HANGUP", it means that:
 - in blocking mode, the call will block until the condition no longer applies
 - in non-blocking mode, the call will end up with appropriate "again" error (as enumerated below)

If the function makes a  "RESUME", it means that it has done something it was intended to do, and in blocking mode it exits at all. In case of non-blocking mode it would exit immediately anyway, but in this case with success.

I think better would be to describe these terms both at one for particular mode (separately for blocking and non-blocking).

Author: How about this:

**HANGUP**: When a function "HANGS UP" this means that:

  - In *blocking* mode the call will block until the condition no longer applies.
  - In *non-blocking* mode the call will end up with appropriate error. It returns 
  an error from the `MJ_AGAIN`  category (see `SRT_EASYNC*`, `SRT_ETIMEOUT` and 
  `SRT_ECONGEST` symbols from the `SRT_ERRNO` enumeration type.

**RESUME**: When a function "RESUMES" this means that:

  - In *blocking* the function has done something it was intended to do and exits.
  - In *non-blocking* the function has done something and returned a non-error 
  status. It would exit immediately anyway, but in this case successfully.

Note that blocking mode in SRT is separate for sending and receiving, and is set 
by the `SRTO_SNDSYN` and `SRTO_RCVSYN` options respectively.

Reviewer: Not sure how clear the "condition" statement is. I'd rather speak about "readiness".

HANGUP is when the function is not ready to perform the operation. In blocking mode it will block, in non-blocking mode will return an error as per this list.

RESUME is when the function is (or better: becomes) ready to perform the operation and it performs it. In both modes the same way.

Author: ### Terminology

The following terms are used in the description of transmission types:

**HANGUP**: When a function "HANGS UP" this means that it is not ready to perform 
an operation:

  - In *blocking* mode the call will block until the condition no longer applies.
  - In *non-blocking* mode the call will return the appropriate error from the 
  `MJ_AGAIN`  category (see `SRT_EASYNC*`, `SRT_ETIMEOUT` and   `SRT_ECONGEST` 
  symbols from the `SRT_ERRNO` enumeration type.

**RESUME**: When a function "RESUMES" this means that it has become ready to perform 
an operation:

  - In *blocking* mode the function has done something it was intended to do and exits.
  - In *non-blocking* mode the function has done something and returned a non-error 
  status. It would exit immediately anyway, but in this case successfully.

In the descriptions further below, these terms denote the readiness of a function 
under certain conditions. Note that blocking mode in SRT is separate for sending 
and receiving, and is set by the `SRTO_SNDSYN` and `SRTO_RCVSYN` options respectively.

> error from the `MJ_AGAIN`  category (see `SRT_EASYNC*`, `SRT_ETIMEOUT` and 
> `SRT_ECONGEST` symbols from the `SRT_ERRNO` enumeration type), if it's in 
> non-blocking mode. 
> 
>   - In blocking mode it will block until the condition that caused the HANGUP no 
>   longer applies, which by definition means the function RESUMES. 
> 
>   - In nonblocking mode, a function RESUMES when the call to it has done something 
>   and returned the non-error status. 
> 
>   - Blocking mode in SRT is separate for sending and receiving, and is set by 
>   the `SRTO_SNDSYN` and `SRTO_RCVSYN` options respectively.

Author: ```suggestion

```

has nothing to do with the "datagram" term in networking terminology, where its 
size is limited to as much it can fit in one MTU. In UDT it is actually a message, 
which may span multiple UDP packets and has clearly defined boundaries. It's rather 
similar to the **SCTP** protocol. Also, in UDP the API functions were strictly bound 
to `DGRAM` or `STREAM` mode: `UDT::send/UDT::recv` were only for `STREAM` and
`UDT::sendmsg/UDT::recvmsg` only for `DGRAM`. In SRT this is changed: all functions 
can be used in all modes, except `srt_sendfile/srt_recvfile`, and how the functions 
actually work is controlled by the `SRTO_MESSAGEAPI` flag.

In message mode, every sending function sends **exactly** as much data as it is 
passed in a single sending function call. The receiver also receives not less than 
**exactly** the number of bytes that was sent (although every message may have a 
different size). Every message may also have extra parameters:

  - **TTL** defines how much time (in ms) the message should wait in the sending 
  buffer for the opportunity to be picked up by the sender thread and sent over 

>   the network; otherwise it is dropped.

Reviewer: There's one thing nonprecise here: The packet is ALWAYS sent for the first time. The TTL parameter sets only a time tolerance for a case when the packet was lost and had to be retransmitted. After that time the retransmission is given up, and if the receiver peer still requests this retransmission, it will be denied by DROPREQ message. I don't think it should be described here in such details, but it should be clearly stated that TTL does not concern regular first-time sending of the packet - which is always done regardless of any conditions - but only if any of the packets carrying this message were lost.

>   - **TTL** defines how much time (in ms) the message should wait in the sending 
>   buffer for the opportunity to be picked up by the sender thread and sent over 
>   the network; otherwise it is dropped.

Author: ```suggestion
  - **TTL** defines how much time (in ms) the message should wait in the sending 
  buffer for the opportunity to be picked up by the sender thread and sent over 
  the network; otherwise it is dropped. Note that this TTL only applies to packets that 
  have been lost and should be retransmitted.
```
